File: mcp-dsl-implementation.test.ts
import { describe, test, expect } from "bun:test";
import { parseMCPDSL, MCPDSLLexer, MCPDSLParser, MCPDSLCompiler } from "./mcp-dsl-implementation";

describe("MCP-DSL Parser & Compiler - Real MCP Spec Tests", () => {
  test("should parse simple ping", () => {
    const dsl = `> ping#2`;
    const result = parseMCPDSL(dsl);

    expect(result.jsonrpc).toBe("2.0");
    expect(result.id).toBe(2);
    expect(result.method).toBe("ping");
  });

  test("should parse initialize request", () => {
    const dsl = `> initialize#1 {
      v: "2025-03-26"
    }`;

    const result = parseMCPDSL(dsl);

    expect(result.jsonrpc).toBe("2.0");
    expect(result.id).toBe(1);
    expect(result.method).toBe("initialize");
    expect(result.params.protocolVersion).toBe("2025-03-26");
  });

  test("should compile Tool definition", () => {
    const dsl = `T search {
      desc: "Search tool"
      in: {query: str!}
    }`;

    const lexer = new MCPDSLLexer(dsl);
    const tokens = lexer.tokenize();
    const parser = new MCPDSLParser(tokens);
    const ast = parser.parse();
    const compiler = new MCPDSLCompiler();
    const result = compiler.compile(ast);

    expect(result.name).toBe("search");
    expect(result.description).toBe("Search tool");
    expect(result.inputSchema).toBeDefined();
  });

  test("should parse notification without id", () => {
    const dsl = `! initialized`;
    const result = parseMCPDSL(dsl);

    expect(result.jsonrpc).toBe("2.0");
    expect(result.method).toBe("initialized");
    expect(result.id).toBeUndefined();
  });

  test("should parse tools/call request", () => {
    const dsl = `> tools/call#4 {
      name: "get_weather"
      args: {location: "New York"}
    }`;

    const result = parseMCPDSL(dsl);

    expect(result.jsonrpc).toBe("2.0");
    expect(result.id).toBe(4);
    expect(result.method).toBe("tools/call");
    expect(result.params.name).toBe("get_weather");
    expect(result.params.arguments.location).toBe("New York");
  });

  test("should parse resources/read request", () => {
    const dsl = `> resources/read#6 {
      uri: "file:///project/src/main.rs"
    }`;

    const result = parseMCPDSL(dsl);

    expect(result.jsonrpc).toBe("2.0");
    expect(result.id).toBe(6);
    expect(result.method).toBe("resources/read");
    expect(result.params.uri).toBe("file:///project/src/main.rs");
  });

  test("should parse error response", () => {
    const dsl = `x #10 -32601:"Method not found"`;

    const result = parseMCPDSL(dsl);

    expect(result.jsonrpc).toBe("2.0");
    expect(result.id).toBe(10);
    expect(result.error).toBeDefined();
    expect(result.error.code).toBe(-32601);
    expect(result.error.message).toBe("Method not found");
  });

  test("should compile Resource definition with annotations", () => {
    const dsl = `R main_file {
      uri: "file:///project/src/main.rs"
      mime: "text/x-rust"
      desc: "Primary application entry point"
      @priority: 1.0
    }`;

    const lexer = new MCPDSLLexer(dsl);
    const tokens = lexer.tokenize();
    const parser = new MCPDSLParser(tokens);
    const ast = parser.parse();
    const compiler = new MCPDSLCompiler();
    const result = compiler.compile(ast);

    expect(result.name).toBe("main_file");
    expect(result.uri).toBe("file:///project/src/main.rs");
    expect(result.mimeType).toBe("text/x-rust");
    expect(result.description).toBe("Primary application entry point");
    expect(result.annotations).toBeDefined();
    expect(result.annotations.priority).toBe(1.0);
  });

  test("should handle required fields in tool schema", () => {
    const dsl = `T weather {
      in: {
        location: str!
        days: int
      }
    }`;

    const lexer = new MCPDSLLexer(dsl);
    const tokens = lexer.tokenize();
    const parser = new MCPDSLParser(tokens);
    const ast = parser.parse();
    const compiler = new MCPDSLCompiler();
    const result = compiler.compile(ast);

    expect(result.inputSchema.required).toContain("location");
    expect(result.inputSchema.required).not.toContain("days");
  });

  test("should map DSL types to JSON Schema types", () => {
    const dsl = `T test_types {
      in: {
        text: str!
        count: int!
        amount: num!
        active: bool!
      }
    }`;

    const lexer = new MCPDSLLexer(dsl);
    const tokens = lexer.tokenize();
    const parser = new MCPDSLParser(tokens);
    const ast = parser.parse();
    const compiler = new MCPDSLCompiler();
    const result = compiler.compile(ast);

    expect(result.inputSchema.properties.text.type).toBe("string");
    expect(result.inputSchema.properties.count.type).toBe("integer");
    expect(result.inputSchema.properties.amount.type).toBe("number");
    expect(result.inputSchema.properties.active.type).toBe("boolean");
  });

  test("should compile @readonly annotation on tools", () => {
    const dsl = `T search {
      desc: "Search tool"
      in: {query: str!}
      @readonly
    }`;

    const lexer = new MCPDSLLexer(dsl);
    const tokens = lexer.tokenize();
    const parser = new MCPDSLParser(tokens);
    const ast = parser.parse();
    const compiler = new MCPDSLCompiler();
    const result = compiler.compile(ast);

    expect(result.annotations).toBeDefined();
    expect(result.annotations.readOnlyHint).toBe(true);
  });

  test("MCP spec: responses must have result, not error", () => {
    const dsl = `< #100 {status: "ok"}`;
    const result = parseMCPDSL(dsl);

    expect(result.id).toBe(100);
    expect(result.result).toBeDefined();
    expect(result.error).toBeUndefined();
    expect(result.method).toBeUndefined();
  });

  test("MCP spec: errors must have error, not result", () => {
    const dsl = `x #100 -32601:"Method not found"`;
    const result = parseMCPDSL(dsl);

    expect(result.id).toBe(100);
    expect(result.error).toBeDefined();
    expect(result.result).toBeUndefined();
    expect(result.method).toBeUndefined();
  });

  test("MCP spec: protocol version 2025-03-26", () => {
    const dsl = `> initialize#1 {v: "2025-03-26"}`;
    const result = parseMCPDSL(dsl);

    expect(result.params.protocolVersion).toBe("2025-03-26");
  });

  test("Token efficiency: DSL should be more compact than JSON", () => {
    const dsl = `> tools/call#42 {name: "search", args: {q: "test"}}`;

    const jsonEquivalent = JSON.stringify({
      jsonrpc: "2.0",
      id: 42,
      method: "tools/call",
      params: {
        name: "search",
        arguments: { q: "test" }
      }
    });

    // DSL should be significantly shorter
    expect(dsl.length).toBeLessThan(jsonEquivalent.length);
  });
});

File: mcp_dsl_implementation_test.nim
import std/[unittest, json]
import mcp_dsl_implementation

suite "MCP-DSL Parser & Compiler - Real MCP Spec Tests":
  test "should parse simple ping":
    let dsl = "> ping#2"
    let result = parseMCPDSL(dsl)

    check result["jsonrpc"].getStr == "2.0"
    check result["id"].getInt == 2
    check result["method"].getStr == "ping"

  test "should parse initialize request":
    let dsl = """
> initialize#1 {
  v: "2025-03-26"
}
    """

    let result = parseMCPDSL(dsl)

    check result["jsonrpc"].getStr == "2.0"
    check result["id"].getInt == 1
    check result["method"].getStr == "initialize"
    check result["params"]["protocolVersion"].getStr == "2025-03-26"

  test "should compile Tool definition":
    let dsl = """
T search {
  desc: "Search tool"
  in: {query: str!}
}
    """

    let result = parseMCPDSL(dsl)

    check result["name"].getStr == "search"
    check result["description"].getStr == "Search tool"
    check result.hasKey("inputSchema")

  test "should parse notification without id":
    let dsl = "! initialized"
    let result = parseMCPDSL(dsl)

    check result["jsonrpc"].getStr == "2.0"
    check result["method"].getStr == "initialized"
    check not result.hasKey("id")

  test "should parse tools/call request":
    let dsl = """
> tools/call#4 {
  name: "get_weather"
  args: {location: "New York"}
}
    """

    let result = parseMCPDSL(dsl)

    check result["jsonrpc"].getStr == "2.0"
    check result["id"].getInt == 4
    check result["method"].getStr == "tools/call"
    check result["params"]["name"].getStr == "get_weather"
    check result["params"]["arguments"]["location"].getStr == "New York"

  test "should parse resources/read request":
    let dsl = """
> resources/read#6 {
  uri: "file:///project/src/main.rs"
}
    """

    let result = parseMCPDSL(dsl)

    check result["jsonrpc"].getStr == "2.0"
    check result["id"].getInt == 6
    check result["method"].getStr == "resources/read"
    check result["params"]["uri"].getStr == "file:///project/src/main.rs"

  test "should parse error response":
    let dsl = """x #10 -32601:"Method not found""""

    let result = parseMCPDSL(dsl)

    check result["jsonrpc"].getStr == "2.0"
    check result["id"].getInt == 10
    check result.hasKey("error")
    check result["error"]["code"].getInt == -32601
    check result["error"]["message"].getStr == "Method not found"

  test "should compile Resource definition with annotations":
    let dsl = """
R main_file {
  uri: "file:///project/src/main.rs"
  mime: "text/x-rust"
  desc: "Primary application entry point"
  @priority: 1.0
}
    """

    let result = parseMCPDSL(dsl)

    check result["name"].getStr == "main_file"
    check result["uri"].getStr == "file:///project/src/main.rs"
    check result["mimeType"].getStr == "text/x-rust"
    check result["description"].getStr == "Primary application entry point"
    check result.hasKey("annotations")
    check result["annotations"]["priority"].getFloat == 1.0

  test "should handle required fields in tool schema":
    let dsl = """
T weather {
  in: {
    location: str!
    days: int
  }
}
    """

    let result = parseMCPDSL(dsl)

    var hasLocation = false
    var hasDays = false
    for item in result["inputSchema"]["required"]:
      if item.getStr == "location":
        hasLocation = true
      elif item.getStr == "days":
        hasDays = true

    check hasLocation
    check not hasDays

  test "should map DSL types to JSON Schema types":
    let dsl = """
T test_types {
  in: {
    text: str!
    count: int!
    amount: num!
    active: bool!
  }
}
    """

    let result = parseMCPDSL(dsl)

    check result["inputSchema"]["properties"]["text"]["type"].getStr == "string"
    check result["inputSchema"]["properties"]["count"]["type"].getStr == "integer"
    check result["inputSchema"]["properties"]["amount"]["type"].getStr == "number"
    check result["inputSchema"]["properties"]["active"]["type"].getStr == "boolean"

  test "should compile @readonly annotation on tools":
    let dsl = """
T search {
  desc: "Search tool"
  in: {query: str!}
  @readonly
}
    """

    let result = parseMCPDSL(dsl)

    check result.hasKey("annotations")
    check result["annotations"]["readOnlyHint"].getBool == true

  test "MCP spec: responses must have result, not error":
    let dsl = """< #100 {status: "ok"}"""
    let result = parseMCPDSL(dsl)

    check result["id"].getInt == 100
    check result.hasKey("result")
    check not result.hasKey("error")
    check not result.hasKey("method")

  test "MCP spec: errors must have error, not result":
    let dsl = """x #100 -32601:"Method not found""""
    let result = parseMCPDSL(dsl)

    check result["id"].getInt == 100
    check result.hasKey("error")
    check not result.hasKey("result")
    check not result.hasKey("method")

  test "MCP spec: protocol version 2025-03-26":
    let dsl = """> initialize#1 {v: "2025-03-26"}"""
    let result = parseMCPDSL(dsl)

    check result["params"]["protocolVersion"].getStr == "2025-03-26"

  test "Token efficiency: DSL should be more compact than JSON":
    let dsl = """> tools/call#42 {name: "search", args: {q: "test"}}"""

    let jsonEquivalent = $(%*{
      "jsonrpc": "2.0",
      "id": 42,
      "method": "tools/call",
      "params": {
        "name": "search",
        "arguments": {"q": "test"}
      }
    })

    # DSL should be significantly shorter
    check dsl.len < jsonEquivalent.len
